x/merkledb/proof.go

```diff
@@ -503,24 +495,29 @@ func verifyAllChangeProofKeyValuesPresent(
 }
 
 func (proof *ChangeProof) Empty() bool {
-	return len(proof.KeyValues) == 0 && len(proof.DeletedKeys) == 0 &&
+	return len(proof.KeyChanges) == 0 &&
 		len(proof.StartProof) == 0 && len(proof.EndProof) == 0
 }
 
-// Returns the largest key in [proof.KeyValues] and [proof.DeletedKeys].
-// If there are no keys in the proof, returns [end].
-func (proof *ChangeProof) getLargestKey(end []byte) []byte {
-	largestKey := end
-	if len(proof.KeyValues) > 0 {
-		largestKey = proof.KeyValues[len(proof.KeyValues)-1].Key
-	}
-	if len(proof.DeletedKeys) > 0 {
-		lastDeleted := proof.DeletedKeys[len(proof.DeletedKeys)-1]
-		if bytes.Compare(lastDeleted, largestKey) > 0 || len(proof.KeyValues) == 0 {
-			largestKey = lastDeleted
+// Returns nil iff both hold:
+// 1. [kvs] is sorted by key in increasing order.
+// 2. All keys in [kvs] are in the range [start, end].
+// If [start] is nil, there is no lower bound on acceptable keys.
+// If [end] is nil, there is no upper bound on acceptable keys.
+// If [kvs] is empty, returns nil.
+func verifyKeyChanges(kvs []KeyChange, start, end []byte) error {
+	hasLowerBound := len(start) > 0
+	hasUpperBound := len(end) > 0
+	for i := 0; i < len(kvs); i++ {
+		if i < len(kvs)-1 && bytes.Compare(kvs[i].Key, kvs[i+1].Key) >= 0 {
+			return ErrNonIncreasingValues
+		}
+		if (hasLowerBound && bytes.Compare(kvs[i].Key, start) < 0) ||
+			(hasUpperBound && bytes.Compare(kvs[i].Key, end) > 0) {
+			return ErrStateFromOutsideOfRange
```

Comment:
We should only need to compare `kvs[0]` against `start` and `kvs[-1]` against `end`.